lib/filterx/func-set-fields.c

```diff
@@ -127,11 +129,89 @@ typedef struct FilterXFunctionSetFields_
   GArray *fields;
 } FilterXFunctionSetFields;
 
+static gboolean
+_set_with_fallbacks(FilterXObject *dict_unwrapped, FilterXObject *key, GPtrArray *values)
+{
+  gboolean changed = FALSE;
+
+  if (!values || values->len == 0)
+    return changed;
+
+  for (guint i = 0; i < values->len; i++)
+    {
+      FilterXExpr *value = g_ptr_array_index(values, i);
+      FilterXObject *value_obj = filterx_expr_eval(value);
+      if (!value_obj)
+        {
+          filterx_eval_clear_errors();
+          continue;
+        }
+
+      if (filterx_object_extract_null(value_obj))
+        {
+          filterx_object_unref(value_obj);
+          continue;
+        }
+
+      FilterXObject *value_obj_cloned = filterx_object_clone(value_obj);
+      filterx_object_unref(value_obj);
+
+      if (!filterx_object_set_subscript(dict_unwrapped, key, &value_obj_cloned))
```

Comment:
You are using the FilterXObject interface on `dict_unwrapped` everywhere, so you don't actually need `filterx_ref_unwrap_rw()`.

A readonly local unwrap is enough to check the type and report errors early, everything else will work transparently (and we would avoid a copy as well in case someone somehow manages to produce a set_fields() call that wouldn't modify the dict :D)

Reply:
Ah, I see! Thanks, updated.